staging/src/k8s.io/client-go/openapi3/root.go

```diff
@@ -0,0 +1,174 @@
+	GVSchemasMap(gv schema.GroupVersion) (map[string]interface{}, error)
+}
+
+// openapi3Root implements the Root interface, and encapsulates the
+// fields to retrieve, store the parsed OpenAPI V3 root document.
+type openapi3Root struct {
+	// OpenAPI client to retrieve the OpenAPI V3 documents.
+	client openapi.Client
+}
+
+// Validate openapi3Root implements the Root interface.
+var _ Root = &openapi3Root{}
+
+// NewRoot returns a structure implementing the Root interface,
+// created with the passed rest client.
+func NewRoot(client openapi.Client) Root {
+	return &openapi3Root{client: client}
+}
+
+// GroupVersions returns all the GroupVersions for which the OpenAPI
+// V3 root document has a secondary OpenAPI V3 url. Returns an error
+// if there is a problem retrieving or parsing the OpenAPI V3 root
+// document.
+func (r *openapi3Root) GroupVersions() ([]schema.GroupVersion, error) {
+	paths, err := r.client.Paths()
+	if err != nil {
+		return nil, err
+	}
+	// Example GroupVersion API path: "/apis/apps/v1"
+	gvSchemas := make([]schema.GroupVersion, len(paths))
+	for gvAPIPath := range paths {
+		gv, err := pathToGroupVersion(gvAPIPath)
+		if err != nil {
+			return nil, err
+		}
+		gvSchemas = append(gvSchemas, gv)
+	}
+	return gvSchemas, nil
+}
+
+// GVSchemas returns the schemas for all the resources in a
+// GroupVersion as a pointer to a spec3.OpenAPI struct.
+// Returns an error for problems retrieving or parsing the root
+// document or GroupVersion OpenAPI V3 document.
+func (r *openapi3Root) GVSchemas(gv schema.GroupVersion) (*spec3.OpenAPI, error) {
+	openAPISchemaBytes, err := r.retrieveGVBytes(gv)
+	if err != nil {
+		return nil, err
+	}
+	// Unmarshal the downloaded Group/Version bytes into the spec3.OpenAPI struct.
+	var parsedV3Schema spec3.OpenAPI
+	err = json.Unmarshal(openAPISchemaBytes, &parsedV3Schema)
+	return &parsedV3Schema, err
+}
+
+// GVSchemasMap returns the schemacs for all the resources in a
+// GroupVersion as unstructured bytes. Returns an error for
+// problems retrieving or parsing the root or GroupVersion
+// OpenAPI V3 documents.
+func (r *openapi3Root) GVSchemasMap(gv schema.GroupVersion) (map[string]interface{}, error) {
+	gvOpenAPIBytes, err := r.retrieveGVBytes(gv)
+	if err != nil {
+		return nil, err
+	}
+	var gvMap map[string]interface{}
+	err = json.Unmarshal(gvOpenAPIBytes, &gvMap)
+	return gvMap, err
+}
+
+// retrieveGVBytes returns the schema for a passed GroupVersion as an
+// unstructured slice of bytes or an error if one occurs.
+func (r *openapi3Root) retrieveGVBytes(gv schema.GroupVersion) ([]byte, error) {
+	paths, err := r.client.Paths()
+	if err != nil {
+		return nil, err
+	}
+	apiPath := gvToAPIPath(gv)
+	gvOpenAPI, found := paths[apiPath]
+	if !found {
+		return nil, fmt.Errorf("GroupVersion (%s) not found in OpenAPI V3 root document", gv)
+	}
+	return gvOpenAPI.Schema(runtime.ContentTypeJSON)
+}
+
+// gvToAPIPath maps the passed GroupVersion to a relative api
```

Comment:
Hard to believe that there's nothing in apimachinery that implements these two methods that already?